source/sliding_puzzle.py

```diff
@@ -124,7 +139,61 @@ def _join(self, row: List[PuzzlePiece]) -> List[str]:
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
         output = []
-        for row in self.pieces:
+        for row in self.rows:
             output.extend(self._join(row))
 
         return '\n'.join(output)
+
+    def shuffle(self) -> None:
+        """Shuffle the puzzle by randomizing the order of the pieces"""
+        for row in self.rows:
+            random.shuffle(row)
+        random.shuffle(self.rows)
+        # Clear a random piece
+        random.choice(random.choice(self.rows)).clear()
+
+    def _get_empty_piece_position(self) -> PiecePosition:
+        x, y, index = [
+            (x, y, piece.index)
+            for x, row in enumerate(self.rows)
+            for y, piece in enumerate(row)
+            if piece.empty
+        ][0]
+        return PiecePosition(x=x, y=y, index=index)
+
+    def move_up(self) -> None:
+        """Move the piece above the empty piece down"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the first row
+        if empty_pos.x == 0:
+            return
+        self._swap_pieces(x1=empty_pos.x, y1=empty_pos.y, x2=empty_pos.x - 1, y2=empty_pos.y)
+
+    def move_down(self) -> None:
+        """Move the piece below the empty piece up"""
+        empty_pos = self._get_empty_piece_position()
+        # return if empty piece is on the last row
+        if empty_pos.x == len(self.rows[0]) - 1:
```

Comment:
This one should just be `len(self.rows) - 1`

Reply:
that's move_down right, actually this one is correct, but it made me realize that there is an issue with move_rigth, ```len(self.rows)``` represent length of columns because it is the number of rows in the list

Reply:
> that's move_down right, actually this one is correct, but it made me realize that there is an issue with move_rigth, `len(self.rows)` represent length of columns because it is the number of rows in the list

ok, I confused myself there, you were rigth

Reply:
@Bluenix2 we are both wrong on this one, the first solution based on the index was easier to undestand, we should have stick to that

```
[
    ['a', 'b'. 'c'],
    ['d', 'e', 'f']
]
```
```x``` is the horizontal line, ```y``` is the vertical one,
every element on the first row have a ```y == 0``` not the x value, the ```x``` on the first line are (0, 1, 2)
In this example ```len(self.rows) - 1``` is ```1```,  we can't write ```empty_pos.x == len(self.rows) - 1``` to indicate that we are on the last row, because only the ```y``` value don't change on the last row, on the last row ```x``` have three value again (0, 1, 2)

I'll try to fix all the functions according to this reasoning, but if it's still confusing I prefer index based checking, it was much easier to understand

Reply:
This is what I've tried to explain in this comment: https://github.com/Istalantar/SCJ-2021-classic-clownfish/pull/4#discussion_r669521172

Going back to a coordinate system, X is horizontal.. both `'b'` and 'e' have the X coordinate 1. `'d'` has the Y coordinate 1 and X coordinate 0. *Confusing enough* this becomes rows[1][0] (`[y][x]`).